service/misc_services.cc

```diff
@@ -204,8 +204,10 @@ future<lowres_clock::duration> cache_hitrate_calculator::recalculate_hitrates()
             llogger.debug("Send CACHE_HITRATES update max_diff={}, published_nr={}", _diff, _published_nr);
             ++_published_nr;
             _published_time = now;
-            return _gossiper.add_local_application_state(gms::application_state::CACHE_HITRATES,
-                    gms::versioned_value::cache_hitrates(_gstate)).then([recalculate_duration] {
+            return container().invoke_on(0, [&gstate = _gstate] (cache_hitrate_calculator& self) {
```

Comment:
`recalculate_hitrates()` is called sequentially, right? So `_gstate` cannot be modified concurrently with this.

Reply:
Yes, there is at most one `recalculate_hitrates()` running at a time:

- `run_on` arms a non-periodic local timer,
- timer executes `recalculate_timer()`, which executes `recalculate_hitrates()` and schedules a continuation that calls `run_on` after `recalculate_hitrates()` completes,

In `main.cc`, there is a single `run_on` call which starts the loop described above.